### PR TITLE
Remove const cast - MERGE FIRST

### PR DESCRIPTION
**Pull request by TomasPuverle**
This is the first of several PRs to clean up the code and express more invariants at the type  level.
Specifically, I will be focusing on replacing pointers with references where a nullptr is not possible; and on improving const-correctness.

This PR improves const-correctness for the AST visitor framework.  I did not want to touch the RAM code visitation because I know @azreika is working on it and I didn't want to cause merge issues.  In any case, all const_casts<> inside of AST visitors are now gone.

The other improvement is that the visitor lambdas no longer go through `std::function`, effectively removing one level of virtual dispatch, which should improve performance of compilation a bit.

Finally, as part of this work, I replaced `IterDerefWrapper` with a more generic `TransformIterator`, which simply applies a function to each returned element.  `IterDerefWrapper` is now just a special case of this.

### CODE REVIEW

**Comment by azreika** [comment]
Just a heads up that the team is on break right now, so reviews might be a little slow!

File: src/ast/Aggregator.h
Comment (TomasPuverle) [reply src/ast/Aggregator.h:78]: Correct!  I obviously didn't want to change the names but I agree that `getChildren` is a better name for the interface.  I'm happy to rename `getChildNodesImpl` but not keen on `registerChildren`.  `getChildrenImpl` would be the idiomatic way of naming based on the `public: foo()`/`private: virtual fooImpl()` idiom.  Still, I will defer to your call.

**Comment by b-scholz** [comment]
I agree that `registerChildren()` might not be the best choice if we cannot implement a type-system that automatically registers children. (NB: I still haven't given up on this plan but let's postpone this idea for the time being). 

Assuming we cannot find a type system that can automatically register children. In this scenario, I find the name choice  `getChildrenImpl()` for the IR problematic since `getChildren()` would never be used in any of the client code for AST/RAM, and would only show up in the visitor code (which is a very small code portion).   Note that the client code includes analyses, transformers, ast-to-ram translator, etc.

However, the AST-IR and RAM-IR classes are quite prominent, i.e., new potential developers will study the IR extensively, and might get confused about the suffix `Impl`. They might want to understand what is the real `getChildren()` but its meaning is insignificant since it is only a helper for the visitor code.  

For the sake of readability of the IR, I would prefer to keep the name `getChildNode()` (or prefer `getChildren()`) in AST and RAM, and use another name for the interface that provides const and non-const access for visitors. I would suggest names such as `getChildrenAdapter()` etc. since they are only used in the visitor code. 

As discussed, I would also prefer simple names for the visitors (like `visit()` and `search()`) and it is great that you simplified the code for the visitors.  Note that we haven't touched the visitors for years.  Great work!

**Comment by b-scholz** [comment]
I had a second thought. 
Let's do all the merges and let's discuss the renaming after all merges are in.

**Review by b-scholz — APPROVED** [review]

**Comment by TomasPuverle** [comment]
Hi,

Thanks for your reply and feedback.  I'm happy about your suggestion to do all the renames after the merge(s).

I am, however, not entirely clear on what you mean by "I find the name choice getChildrenImpl() for the IR problematic since getChildren() would never be used in any of the client code for AST/RAM".  
In my mind, no external client (other than a few derived classes) of a `Node` would/should ever call `getChildrenImpl` (hence the name `Impl` and the restricted accessibility of the function).  So I am not sure what you mean by the "clients" of the code.  Every user would go through `getChildren` in the base class `ast::Node`, which then forwards to (and has the option to add additional processing) before/after the call to `getChildrenImpl`.  
As another example of this, please note that I did the same thing with `clone` (in `ast::Node` and `ast::Transformer`).  `ast::Node::clone` now *always* returns an `Own<>`, meaning that you can't accidentally forget to delete the new pointer.  However, since `Own<A>` and `Own<B>` are not covariant, the previous "raw pointer returning" `clone` is now private and called `cloneImpl` and cannot be accessed by the clients of `ast::Node` - only by `Node` itself.  This ensures that it wraps the returned raw pointer in `Own<>` before returning it.

I just want to clarify that I don't have a strong naming preference but before deciding on names, I want to ensure we're on the same page wrt the (re-)design itself.

Finally, if the "automatic detection" of the child nodes is important, I am happy to think about it some more.  I am currently taking a break from continuing the cleanups until after the merges are done, and working on the Python interface instead.  One idea that immediately springs to mind is the possibility of having a nested "struct Children", which may be amenable to automatic "registration"/metaprogramming.  How would you feel about such a change?

Thanks,

Tom

**Pull request merged**
